### PR TITLE
feature/update-tuner-prepare-data-generator

### PR DESCRIPTION
**Pull request by uduse**
Update many many things, including but not limited to:

- add mz.Tuner
- add mz.tuner.callbacks
- add mz.ZooKeeper
- rewrite mz.DataGenerator to be callback-based
    - replace mz.PairDataGenerator by mz.DataGenerator arguments
    - replace mz.HistogramDataGenerator/mz.DPoolDataGenerator by callbacks
- update engine
    - move metrics remapping from evaluation to attribute assignment
    - add mz.engine.BaseModel.evaluate_generator
    - add mz.engine.BaseModel.get_embedding_layer
    - add mz.engine.ParamTable.update
- add mz.losses attributes getters/setters
- fix ANMM bug on classification task
- fix ArcII invalid padding bug
- fix MatchPyramid invalid padding bug
- fix DRMMTKS "top_k could be too large" bug
- add Naive hyper_space
- add embedding_input_dim to preprocessors
- add DSSM/CDSSM preprocessor attributes getters/setters
- update processor units for return type consistency
- mute printing from some doctests
- refactored models unit test
- add hyper_spaces unit test
- add notebook draft

### CODE REVIEW

File: matchzoo/auto/tuner/callbacks/save_model.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import typing
+from pathlib import Path
+import uuid
+
+import matchzoo as mz
+from matchzoo.auto.tuner.callbacks.callback import Callback
+
+
+class SaveModel(Callback):
```
[the remark this answers is not in the document]
Comment (uduse) [reply matchzoo/auto/tuner/callbacks/save_model.py]: That's `fit_kwargs`'s job. e.g. `fit_kwargs=dict(callbacks=[keras.callbacks.ModelCheckpoint()])`

**Comment by uduse** [comment]
More updates.

- move `processor_units` into `preprocessors` as `preprocessors.units`. `processor_units` is too similar to `preprocessors`. this also shows structurally that units are components of preprocessors
- split the units into multiple files so it is easier to extend
- changed all importing related to the old `matchzoo.engine`. the abstraction level of this package makes the initialization of this package extremely hard. changing the way this package should be imported makes future updates to this package and other things much easier
- fix the old `list_available` bug where subclasses of subclasses aren't shown

File: matchzoo/engine/base_model.py [hunk 2]
```diff
@@ -238,7 +251,7 @@ def fit(
 
     def fit_generator(
         self,
-        generator: DataGenerator,
+        generator: matchzoo.DataGenerator,
```
[the remark this answers is not in the document]
Comment (uduse) [reply matchzoo/engine/base_model.py]: I think it's ok. 14 files are using `import matchzoo`. I can make the change if you think it matters.

File: matchzoo/metrics/discounted_cumulative_gain.py [hunk 3]
```diff
@@ -52,7 +52,7 @@ def __call__(self, y_true: np.array, y_pred: np.array) -> float:
         """
         if self._k <= 0:
             return 0.
-        coupled_pair = engine.sort_and_couple(y_true, y_pred)
+        coupled_pair = sort_and_couple(y_true, y_pred)
```
[the remark this answers is not in the document]
Comment (uduse) [reply matchzoo/metrics/discounted_cumulative_gain.py]: I will handle the merge conflict, no worries.

File: matchzoo/data_generator/data_generator.py [hunk 4]
```diff
@@ -5,65 +5,121 @@
...
+            callbacks = []
 
-        :param data_pack: a :class:`DataPack` object.
-        :param batch_size: number of instances for each batch
-        :param shuffle: a bool variable to determine whether choose samples
-        randomly
-        """
-        self._data_pack = data_pack
+        if mode not in ('point', 'pair', 'list'):
+            raise ValueError
```
[the remark this answers is not in the document]
Comment (uduse) [reply matchzoo/data_generator/data_generator.py]: Found 18 raises without information, I could fix all those in a later patch.

File: matchzoo/engine/base_model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (faneshion) [reply matchzoo/engine/base_model.py]: I think `mz` is better. In this way, we can cultivate users' or developers' behavior!
